servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherCache.java

```diff
@@ -0,0 +1,261 @@
+     * consumers of a specific key.
+     *
+     * @param minSubscribers The upstream subscribe operation will not happen until after this many {@link Subscriber}
+     * subscribe to the return value.
+     * @param <K> a key type suitable for use as a {@link Map} key.
+     * @param <T> the type of the {@link Publisher} contained in the cache.
+     * @return a new PublisherCache that will wrap cached values with multicast operator.
+     */
+    public static <K, T> PublisherCache<K, T> multicast(final int minSubscribers) {
+        return new PublisherCache<>(MulticastStrategy.wrapMulticast(minSubscribers));
+    }
+
+    /**
+     * Create a new PublisherCache where the cached publishers will be configured for multicast for all
+     * consumers of a specific key.
+     *
+     * @param minSubscribers The upstream subscribe operation will not happen until after this many {@link Subscriber}
+     * subscribe to the return value.
+     * @param queueLimit The number of elements which will be queued for each {@link Subscriber} in order to compensate
+     * for unequal demand.
+     * @param <K> a key type suitable for use as a {@link Map} key.
+     * @param <T> the type of the {@link Publisher} contained in the cache.
+     * @return aa new PublisherCache that will wrap cached values with multicast operator.
+     */
+    public static <K, T> PublisherCache<K, T> multicast(final int minSubscribers, final int queueLimit) {
+        return new PublisherCache<>(MulticastStrategy.wrapMulticast(minSubscribers, queueLimit));
+    }
+
+    /**
+     * Create a new PublisherCache where the cacehed publishers will be configured for replay given the privided
+     * ReplayStrategy.
+     *
+     * @param replayStrategy a {@link ReplayStrategy} that determines the replay behavior and history retention logic.
+     * @param <K> a key type suitable for use as a {@link Map} key.
+     * @param <T> the type of the {@link Publisher} contained in the cache.
+     * @return a new PublisherCache that will wrap cached values with replay operator.
+     */
+    public static <K, T> PublisherCache<K, T> replay(ReplayStrategy<T> replayStrategy) {
+        return new PublisherCache<>(MulticastStrategy.wrapReplay(replayStrategy));
+    }
+
+    /**
+     * Retrieve the value for the given key, if no value exists in the cache it will be synchronously created.
+     *
+     * @param key a key corresponding to the requested {@link Publisher}.
+     * @param publisherSupplier if the key does not exist in the cache, used to create a new publisher.
+     * @return a new {@link Publisher} from the publisherSupplier if not contained in the cache, otherwise
+     * the cached publisher.
+     */
+    public Publisher<T> get(final K key, final Function<K, Publisher<T>> publisherSupplier) {
+        return Publisher.defer(() -> {
+            synchronized (publisherCache) {
```

Comment:
`synchronized` has been used in the past bcz it doesn't require additional allocations. however loom fibers don't support `synchronized`, should we use `Lock` objects instead?